ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/MLSdkAsyncHttpResponseHandler.java

```diff
@@ -137,10 +138,25 @@ private void processResponse(
     private void reOrderTensorResponses(Map<Integer, ModelTensors> tensorOutputs) {
         ModelTensors[] modelTensors = new ModelTensors[tensorOutputs.size()];
         log.debug("Reordered tensor outputs size is {}", tensorOutputs.size());
+        // step 1: reorder batches
         for (Map.Entry<Integer, ModelTensors> entry : tensorOutputs.entrySet()) {
             modelTensors[entry.getKey()] = entry.getValue();
         }
-        actionListener.onResponse(Arrays.asList(modelTensors));
+
+        // step 2: restore to original order as textDocs might be sorted
+        List<Integer> originalOrderIndexes = executionContext.getOriginalOrder();
+        List<ModelTensors> modelTensorsList = Arrays.asList(modelTensors);
+        if (CollectionUtils.isEmpty(originalOrderIndexes)) {
+            actionListener.onResponse(modelTensorsList);
+        } else {
+            // if the originalOrder is not empty, reorder based on it
+            List<ModelTensors> sortedModelTensors = Arrays.asList(new ModelTensors[modelTensorsList.size()]);
+            assert (originalOrderIndexes.size() == modelTensors.length);
```

Comment:
The `originalOrderIndexes`  size is the doc number, while `modelTensors` size is request size. So it would cause some error here. A simple way to reproduce is to try predict a four size input and set connector max batch size to 2. We may need to fetch all tensor and re-create ModelTensors Again.

One more small detail here: What is the list of `modelTensors` size here we want to keep? If we keep the size equal to request number, it may confuse user because actually sort and modelTensors is not the actual remote response (we re-sort it). So maybe keep all tensor into one modelTensors can be a choice.

Reply:
Thanks for catching this, it is a bug, will fix in next revision